### PR TITLE
feat(console): add connectors transfer component

### PR DESCRIPTION
**Pull request by wangsijie**
## Summary

Add component: ConnectorsTransfer:

1. choose and add connectors.
2. sort by drag and drop.

## Linear Issue Reference

LOG-1735

## Testing

截屏2022-04-18 下午5 59 03
截屏2022-04-18 下午5 59 35

### CODE REVIEW

File: packages/console/src/components/Transfer/DragableItem.tsx [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+type DragItemProps = {
+  sortIndex: number;
+  id: string;
+  type: string;
+};
+
+const drapType = 'TransferItem';
+
+const DragableItem = ({ id, children, sortIndex, moveItem }: Props) => {
```
Comment (demonzoo) [inline packages/console/src/components/Transfer/DragableItem.tsx:22]: drag`g`able

File: packages/console/src/components/Transfer/DragableItem.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  children: ReactNode;
+};
+
+type DragItemProps = {
+  sortIndex: number;
+  id: string;
+  type: string;
+};
+
+const drapType = 'TransferItem';
```
Comment (demonzoo) [inline packages/console/src/components/Transfer/DragableItem.tsx:20]: drap => drag?

File: packages/console/src/components/Transfer/DragableItem.tsx (hunk 1 above)
Comment (wangsijie) [reply packages/console/src/components/Transfer/DragableItem.tsx]: updated

File: packages/console/src/components/Transfer/DragableItem.tsx (hunk 2 above)
Comment (wangsijie) [reply packages/console/src/components/Transfer/DragableItem.tsx]: updated

File: packages/console/src/components/Transfer/DraggableItem.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  id: string;
+  type: string;
+};
+
+const dragType = 'TransferItem';
+
+const DraggableItem = ({ id, children, sortIndex, moveItem }: Props) => {
+  const ref = useRef<HTMLDivElement>(null);
+  const { setIsDragging } = useContext(DragDropContext);
+  const [{ handlerId }, drop] = useDrop<DragItemProps, void, { handlerId: Identifier | undefined }>(
```
Comment (demonzoo) [inline packages/console/src/components/Transfer/DraggableItem.tsx:25]: Nullable

File: packages/console/src/components/Transfer/index.tsx [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        <div className={styles.box}>
+          <div className={styles.title}>
+            {selectedItems.length} {t('general.added')}
+          </div>
+          <div className={styles.items}>
+            <DragDropProvider>
+              {selectedItems.map(({ value, title }, index) => (
+                <DraggableItem key={value} sortIndex={index} id={value} moveItem={onMoveItem}>
+                  <div className={classNames(styles.item, styles.dragable)}>
+                    <Dragable />
```
Comment (demonzoo) [inline packages/console/src/components/Transfer/index.tsx:101]: Draggable

File: packages/console/src/components/Transfer/DraggableItem.tsx [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+        // Note: we're mutating the monitor item here!
+        // Generally it's better to avoid mutations,
+        // but it's good here for the sake of performance
+        // to avoid expensive index searches.
+        // eslint-disable-next-line @silverhand/fp/no-mutation
+        item.sortIndex = hoverIndex;
+      },
+    }
+  );
```
Comment (demonzoo) [inline packages/console/src/components/Transfer/DraggableItem.tsx:82]: Is this a hook config?? Why is this hook so damn hard to initialize? A lot of logics above seems should be provided out of the box, I think

**Comment by demonzoo** [comment]
After reviewing the code, I think another tool called `dnd-kit` is better than `react-dnd`. Check this out:
https://docs.dndkit.com/introduction/getting-started

File: packages/phrases/src/locales/zh-cn.ts [hunk 6]
```diff
@@ -360,8 +361,11 @@ const translation = {
         define_social_methods: 'Define social sign in methods',
         transfer: {
           title: 'Social connectors',
-          added: 'added',
-          footer: 'Not in the list?  Set up more social connectors or go to “Connectors” section.',
+          footer: {
+            not_in_list: 'Not in the list? ',
```
Comment (IceHe) [inline packages/phrases/src/locales/zh-cn.ts:365]: nit: unnecessary whitespace after `?`

File: packages/console/src/components/Transfer/DraggableItem.tsx [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+  const opacity = isDragging ? 0 : 1;
+  drag(drop(ref));
+
+  useEffect(() => {
+    setIsDragging?.(isDragging);
+  }, [setIsDragging, isDragging]);
+
+  return (
+    <div ref={ref} style={{ opacity }} data-handler-id={handlerId}>
```
Comment (IceHe) [inline packages/console/src/components/Transfer/DraggableItem.tsx:102]: nit: `opacity` surrounded by two braces?

File: packages/phrases/src/locales/en.ts [hunk 8]
```diff
@@ -364,8 +365,11 @@ const translation = {
         define_social_methods: 'Define social sign in methods',
         transfer: {
           title: 'Social connectors',
-          added: 'added',
-          footer: 'Not in the list?  Set up more social connectors or go to “Connectors” section.',
+          footer: {
+            not_in_list: 'Not in the list? ',
```
Comment (IceHe) [inline packages/phrases/src/locales/en.ts:369]: nit: unnecessary whitespace after ?

File: packages/console/src/components/Transfer/index.tsx [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  title: ReactNode;
+  value: string;
+};
+
+type Props = {
+  title: string;
+  datasource: TransferItem[];
+  value?: string[];
+  footer?: ReactNode;
+  onChange?: (value: string[]) => void;
```
Comment (IceHe) [inline packages/console/src/components/Transfer/index.tsx:24]: why `onChange` optional?

File: packages/console/src/components/Transfer/DraggableItem.tsx (hunk 5 above)
Comment (wangsijie) [reply packages/console/src/components/Transfer/DraggableItem.tsx:82]: So sad but true, we need to write those code, copied from offical examples.

File: packages/console/src/components/Transfer/index.tsx (hunk 9 above)
Comment (wangsijie) [reply packages/console/src/components/Transfer/index.tsx:24]: Is a common practice, an example: we only need this component to display things (preview or readonly).

File: packages/console/src/components/Transfer/DraggableItem.tsx (hunk 5 above)
Comment (demonzoo) [reply packages/console/src/components/Transfer/DraggableItem.tsx:82]: Well, I'll see if I can refactor it with `dnd-kit`. The API seems way better than this

**Comment by wangsijie** [comment]
@demonzoo As for the final result, I don't see much difference between these these 2 packages, though, `dnd-kit` is cool, maybe we can refactor in the future?

File: packages/phrases/src/locales/zh-cn.ts (hunk 6 above)
Comment (wangsijie) [reply packages/phrases/src/locales/zh-cn.ts]: updated

File: packages/phrases/src/locales/en.ts (hunk 8 above)
Comment (wangsijie) [reply packages/phrases/src/locales/en.ts]: updated

File: packages/console/src/components/Transfer/index.tsx (hunk 4 above)
Comment (wangsijie) [reply packages/console/src/components/Transfer/index.tsx:101]: updated

File: packages/console/src/components/Transfer/DraggableItem.tsx (hunk 3 above)
Comment (wangsijie) [reply packages/console/src/components/Transfer/DraggableItem.tsx]: updated

File: packages/console/src/components/Transfer/index.tsx [hunk 10]
```diff
@@ -0,0 +1,123 @@
+import classNames from 'classnames';
+import React, { ReactNode, useMemo } from 'react';
+import { useTranslation } from 'react-i18next';
+
+import CircleMinus from '@/icons/CircleMinus';
+import CirclePlus from '@/icons/CirclePlus';
+import Dragable from '@/icons/Dragable';
```
Comment (demonzoo) [inline packages/console/src/components/Transfer/index.tsx:7]: Still... Drag`g`able. Please search and replace globally, including file names

**Comment by demonzoo** [comment]
> @demonzoo As for the final result, I don't see much difference between these these 2 packages, though, `dnd-kit` is cool, maybe we can refactor in the future?

Yes, this is exactly what I'm saying. We'll see if we can get the bandwidth to refactor. Or maybe it turns out to be that `dnd-kit` won't serve the purpose here. We don't know it yet...
Anyways, it's good to leave it as it is for now.

File: packages/console/src/components/Transfer/index.tsx (hunk 9 above)
Comment (IceHe) [reply packages/console/src/components/Transfer/index.tsx:24]: nit: it seems that `onChange` is required for the `Transfer` component in most cases. I'm good with both for now.

Comment (wangsijie) [reply packages/console/src/components/Transfer/index.tsx:24]: `Transfer` is a standard form component, `onChange` should be optional.

File: packages/console/src/components/Transfer/index.tsx (hunk 10 above)
Comment (wangsijie) [reply packages/console/src/components/Transfer/index.tsx]: updated, sorry for the typo issues.

**Comment by wangsijie** [comment]
@demonzoo Created a research issue: LOG-2231

**Review by demonzoo — APPROVED** [review]
Looks good!
